ingestion-beam/src/main/java/com/mozilla/telemetry/ContextualServicesReporter.java

```diff
@@ -70,31 +72,65 @@ public static PipelineResult run(ContextualServicesReporterOptions.Parsed option
         .failuresTo(errorCollections) //
         .apply(EmitCounters.of());
 
-    Set<String> aggregatedDocTypes = ImmutableSet.of("topsites-impression");
-    Set<String> perContextIdDocTypes = ImmutableSet.of("topsites-click");
+    Set<String> individualImpressions = ImmutableSet.of("topsites-impression");
+    Set<String> individualClicks = ImmutableSet.of("topsites-click");
+
     Set<String> unionedDocTypes = Stream
-        .concat(aggregatedDocTypes.stream(), perContextIdDocTypes.stream())
+        .concat(individualImpressions.stream(), individualClicks.stream())
         .collect(Collectors.toSet());
 
+    // Perform windowed click counting per context_id, adding a click-status to the reporting URL
+    // if the count passes a threshold.
+    PCollection<SponsoredInteraction> clicksCountedByContextId = requests
+        .apply("FilterPerContextIdDocTypes", Filter.by((interaction) -> individualClicks //
+            .contains(interaction.getDerivedDocumentType())))
+        .apply(LabelSpikes.perContextId(options.getClickSpikeThreshold(),
+            Time.parseDuration(options.getClickSpikeWindowDuration()), TelemetryEventType.CLICK));
+
+    // Perform windowed impression counting per context_id, adding an impression-status to the
+    // reporting URL
+    // if the count passes a threshold.
+    PCollection<SponsoredInteraction> impressionsCountedByContextId = requests
+        .apply("FilterPerContextIdDocTypes", Filter.by((interaction) -> individualImpressions //
+            .contains(interaction.getDerivedDocumentType())))
+        .apply(LabelSpikes.perContextId(options.getImpressionSpikeThreshold(),
+            Time.parseDuration(options.getImpressionSpikeWindowDuration()),
+            TelemetryEventType.IMPRESSION));
+
     // Aggregate impressions.
-    PCollection<SponsoredInteraction> aggregated = requests
-        .apply("FilterAggregatedDocTypes", Filter.by((interaction) -> aggregatedDocTypes //
+    PCollection<SponsoredInteraction> aggregatedGenuineImpressions = requests
```

Comment:
This should aggregate `impressionsCountedByContextId` rather than `requests` which does not have the labeling applied to it yet

Reply:
ty!